pkg/cache/cache.go

```diff
@@ -427,7 +427,7 @@ func defaultOpts(config *rest.Config, opts Options) (Options, error) {
 		byObject.Transform = defaultedConfig.Transform
 		byObject.UnsafeDisableDeepCopy = defaultedConfig.UnsafeDisableDeepCopy
 
-		if byObject.Namespaces == nil {
+		if isNamespaced && byObject.Namespaces == nil {
```

Comment:
Trying to understand the impact of this, does the incorrect defaulting actually result in incorrect behavior?

Reply:
Yes, It panics in the below scenario, when `DefaultNamespaces` namespace is present and in `ByObject` non-namespaced object is present.
```
	mgr, err := manager.New(restConfig, manager.Options{
		Metrics: metricsserver.Options{BindAddress: "0"},
		Cache: cache.Options{
			DefaultNamespaces: map[string]cache.Config{testNamespace.Name: {}},
			ByObject: map[client.Object]cache.ByObject{
				&corev1.Node{}: {
					Label: labels.SelectorFromSet(labels.Set{testID: testRunID}),
				},
			},
		},
	})
```

Reply:
Stack trace of panic - 
```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x20 pc=0x1056d7a8c]

goroutine 1709 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0x140027b3978?})
	/Users/I545932/go/src/github.com/gardener/gardener/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:56 +0xe0
panic({0x107020a40?, 0x108e3c760?})
	/opt/homebrew/opt/go/libexec/src/runtime/panic.go:914 +0x218
sigs.k8s.io/controller-runtime/pkg/cache.(*multiNamespaceCache).GetInformer(0x140032b80c0, {0x10750e250, 0x14000bdf450}, {0x107533070?, 0x14002143900?}, {0x0, 0x0, 0x0})
	/Users/I545932/go/src/github.com/gardener/gardener/vendor/sigs.k8s.io/controller-runtime/pkg/cache/multi_namespace_cache.go:87 +0xcc
sigs.k8s.io/controller-runtime/pkg/cache.(*delegatingByGVKCache).GetInformer(0x14002b43c88?, {0x10750e250, 0x14000bdf450}, {0x107533070?, 0x14002143900?}, {0x0, 0x0, 0x0})
	/Users/I545932/go/src/github.com/gardener/gardener/vendor/sigs.k8s.io/controller-runtime/pkg/cache/delegating_by_gvk_cache.go:60 +0xa0
sigs.k8s.io/controller-runtime/pkg/internal/source.(*Kind).Start.func1.1({0x10750e250?, 0x14000bdf450?})
	/Users/I545932/go/src/github.com/gardener/gardener/vendor/sigs.k8s.io/controller-runtime/pkg/internal/source/kind.go:58 +0x64
k8s.io/apimachinery/pkg/util/wait.loopConditionUntilContext.func1(0x14000bdf450?, {0x10750e250?, 0x14000bdf450?})
	/Users/I545932/go/src/github.com/gardener/gardener/vendor/k8s.io/apimachinery/pkg/util/wait/loop.go:49 +0x58
k8s.io/apimachinery/pkg/util/wait.loopConditionUntilContext({0x10750e250, 0x14000bdf450}, {0x1074fd1c8?, 0x14000d543e0}, 0x1, 0x0, 0x0?)
	/Users/I545932/go/src/github.com/gardener/gardener/vendor/k8s.io/apimachinery/pkg/util/wait/loop.go:50 +0xa0
k8s.io/apimachinery/pkg/util/wait.PollUntilContextCancel({0x10750e250, 0x14000bdf450}, 0x0?, 0x0?, 0x0?)
	/Users/I545932/go/src/github.com/gardener/gardener/vendor/k8s.io/apimachinery/pkg/util/wait/poll.go:33 +0x74
sigs.k8s.io/controller-runtime/pkg/internal/source.(*Kind).Start.func1()
	/Users/I545932/go/src/github.com/gardener/gardener/vendor/sigs.k8s.io/controller-runtime/pkg/internal/source/kind.go:56 +0xb8
created by sigs.k8s.io/controller-runtime/pkg/internal/source.(*Kind).Start in goroutine 1704
	/Users/I545932/go/src/github.com/gardener/gardener/vendor/sigs.k8s.io/controller-runtime/pkg/internal/source/kind.go:48 +0x1c4
```

Reply:
Panics are always wrong. Are you sure this doesn't also panic with namespaced objects? And we will need a regression test to verify that this is fixed and remains fixed.

Reply:
> Are you sure this doesn't also panic with namespaced objects?

I have tested it for namespaced objects, it is fine.

> With the  change in the PR 

It works for both namespaced and non-namespaced objects.

Reply:
I have added test in 26c6cffb07e2b54116a4a2039d78b67cf4983130